Update Allo Piano Plus dac driver

paul-1 opened the pull request:
The current driver is non-functional with kernel 5.4.y.

Issues are related to having 2 identical dac codec drivers on the card.  The current kernels require the drivers to be named as well as unique mixer controls.   Changes are similar to how the new PiFi-40 driver is done.

@allocom (@alloboss2)  Please review.

pelwell commented on sound/soc/bcm/allo-piano-dac-plus.c:
```diff
@@ -276,8 +277,13 @@ static int snd_allo_piano_dual_mode_put(struct snd_kcontrol *kcontrol,
 		list_for_each_entry(kctl, &snd_card_ptr->controls, list) {
-			if (!strncmp(kctl->id.name, "Digital Playback Volume",
-					sizeof(kctl->id.name))) {
+			if (!strncmp(kctl->id.name, "Main Digital Playback Volume",sizeof(kctl->id.name))) {
```
This line looks too long, and if you don't split it after the second comma it needs a space. The same applies to several other lines.

paul-1 replied:
Changed.

pelwell commented on sound/soc/bcm/allo-piano-dac-plus.c:
```diff
@@ -719,12 +753,34 @@ static int snd_allo_piano_dac_init(struct snd_soc_pcm_runtime *rtd)
+	}
+
+	// Remove codec controls
+	for (i = 0; i < 2; i++) {
```
`ARRAY_SIZE(codec_ctl_pfx)`?

paul-1 replied:
Makes sense, I was just copying and pasting from another driver.  But much easier to spot in a couple lines of a patch, rather than the complete driver.  Will make those changes

pelwell commented on sound/soc/bcm/allo-piano-dac-plus.c:
```diff
@@ -719,12 +753,34 @@ static int snd_allo_piano_dac_init(struct snd_soc_pcm_runtime *rtd)
+
+	// Remove codec controls
+	for (i = 0; i < 2; i++) {
+		for (j = 1; j < 12; j++) {  //Leave the Digital Volume Control.
```
`ARRAY_SIZE(codec_ctl_name)`?

pelwell commented on sound/soc/bcm/allo-piano-dac-plus.c:
```diff
@@ -719,12 +753,34 @@ static int snd_allo_piano_dac_init(struct snd_soc_pcm_runtime *rtd)
-			dev_warn(card->dev, "Failed to set volume limit: %d\n",
-				ret);
+		//Set volume limit on both dacs
+		for (i = 0; i < 2; i++) {
```
`ARRAY_SIZE(codec_ctl_pfx)`?

pelwell reviewed (commented):
We expect all contributions to follow the kernel coding rules. The easiest way to verify your patches it to use scripts/checkpatch.pl on them (you don't need to worry about documenting new compatible strings or adding files to MAINTAINERS.

The commit message could also do with a bit more explanation of the changes.

You can wait for feedback from the Allo devs before pushing an update, if you want.

paul-1 committed:
Update Allo Piano Dac Driver for 5.4.y kernels

Add unique names to the individual dac coded drivers
Remove some of the codec controls that are not used.

Signed-off-by: Paul Hermann <paul@picoreplayer.org>

paul-1 committed:
Overlay: Update Allo Piano Plus dac driver for 5.4.y kernels.

Create unique names for the two instances of the codec driver.

Signed-off-by: Paul Hermann <paul@picoreplayer.org>

paul-1 commented:
As many times as I've seen you tell folks to use checkpatch, I should have known better.   I think I've got things fixed up, but lets see what Allo has to say.

pelwell commented:
I'm happy with that. Give Allo a few days to look it over and test it, but if you hear nothing give me a nudge and I'll merge it as is.

alloboss2 commented:
@pelwell @paul-1    Thanks for your help. Allo will do the testing with hardware, 
Testing will take 2-3 days. we will update you.

paul-1 committed:
Update volume controls in Allo Piano Dac Plus

Put control scaling back to 255.
Clean up what master/sub volume controls set in codec.
Remove more unneeded mixer controls.

Signed-off-by: Paul Hermann <paul@picoreplayer.org>

paul-1 commented:
As I was updating mixer integration in our interface, I found some oddities in the volume control.  I think this works much better now.

Once we come to an agreement on the updates, I can squash the commits if needed.

paul-1 commented:
Adding link to open issue.  I had not seen that until today.

#3932

pelwell commented:
Allo, is there any feedback?

alloboss2 commented:
@pelwell  @paul-1 
 Yes We have completed the testing and working fine with latest changes. Driver updates are fine, please proceed.

pelwell commented:
Thanks, @paul-1.

allocom commented:
> 
> 
> @pelwell @paul-1
> Yes We have completed the testing and working fine with latest changes. Driver updates are fine, please proceed.

@pelwell  These commits reflected in 5.4.77 kernel onwards , but its not reflecting in 5.10.y repository, please take care

pelwell commented:
These commits don't apply in 5.10.y - please fix them up.

paul-1 commented:
I can take a look at this.   I'll open a new PR when I get them fixed up.

allocom commented:
> 
> 
> I can take a look at this. I'll open a new PR when I get them fixed up.

@paul-1 
thanks , done below update.  checking the board with below changes.
https://github.com/allocom/linux/commit/b9f4b9e1e991180efb80f52135f7ad8406dd2290

paul-1 commented:
So does that mean you are going to handle it.  That works for me.   Just be aware that branch above does not have the updated overlay in it.  There were 3 commits in this original PR, make sure to apply them all in your update.   Although I think they normally like the overlay update in a separate commit.

ashthespy commented:
@paul-1 Sorry to necrobump - but this driver [was recently brought up](https://community.volumio.org/t/allo-piano-2-1-sub-volume-not-linked-to-master-volume-pi-4b-volumio-3-spotify/52957), and I had a question about the implementation of the `Master playback Volume` mixer.

Is it intentional that in non dual mono mode, only `codec_dais[0]` is being set?
https://github.com/raspberrypi/linux/blob/ec0dcf3064b8ba99f226438214407fcea9870f76/sound/soc/bcm/allo-piano-dac-plus.c#L601-L611

Previously, this wasn't the case? 
https://github.com/raspberrypi/linux/blob/01d5eea0c847d3c337acf00dedcceaa1e5b8a548/sound/soc/bcm/allo-piano-dac-plus.c#L562-L581

To summarise: 

| SHA |`dual_mode`| `codec_dais[0] - L` | `codec_dais[0] - R ` | `codec_dais[1] - L` | `codec_dais[1] - R` | 
| -- | ---- | ---- | ---- |  ---- |  ---- | 
|ec0dcf3064b8ba99f226438214407fcea9870f76| `==1` | `left_val` |  |  | `right_val`|
|01d5eea|`==1` | `left_val` |  |  | `right_val`|
|ec0dcf3064b8ba99f226438214407fcea9870f76|  `!=1` | `left_val` | `right_val` |  | |
|01d5eea|`!=1` |  | `right_val` | `left_val` | |

I was curious and looked up the different registers, and if [I understood the datasheet registers correctly](https://www.ti.com/lit/ds/symlink/pcm5122.pdf?ts=1642598435761), setting [`PCM512x_DIGITAL_VOLUME_1=0x00`](https://github.com/raspberrypi/linux/blob/ec0dcf3064b8ba99f226438214407fcea9870f76/sound/soc/codecs/pcm512x.c#L87) implies independent volume for each channel.

Wouldn't it be then more appropriate to have something like this mapping?
|`dual_mode`| `codec_dais[0] - L` | `codec_dais[0] - R ` | `codec_dais[1] - L` | `codec_dais[1] - R` | 
| ---- | ---- | ---- |  ---- |  ---- | 
|`==1` | `left_val` |  |  | `right_val`|
|`!=1` | `left_val` | `right_val` | `left_val` | `right_val` |


PS: Disclaimer - I have just enough knowledge here to start being dangerous...

paul-1 commented:
This driver will make your head hurt.  If I recall, (2021 seems forever ago :)  The Master volume control really never worked properly in all modes.  So I made a decision to try to make the Master control not make a wild volume swing.   In Dual mode, master will keep Dac0 and Dac1 in volume sync.  But when in non-Dual mode, Master only controls Dac0 and Sub volume controls Dac1

I would imagine the driver could be re-written to provide a true Master Volume.  But it would need to handle relative volume adjustments for each DAC, Since you may not want them at the same volume.  That's not something I want to take on.  Perhaps you can talk with @allocom.

ashthespy commented:
Thanks for the clarification! :-) 

>I would imagine the driver could be re-written to provide a true Master Volume.  But it would need to handle relative volume adjustments for each DAC, Since you may not want them at the same volume.

Could I pick your brain a bit more  -- I don't have this device on hand, but shouldn't that already be halfway implemented via the default [`Digital Playback Volume`](https://github.com/raspberrypi/linux/blob/95d7686066c3261e9104970a87f279afc49a9d64/sound/soc/codecs/pcm512x.c#L438) the codec exposes? Allo's docs seem to indicate this is what was envisioned as well?

<img src="https://community.volumio.org/uploads/default/original/2X/e/e69b8f753d3ec020420841701cc41e59e46b2a82.png" alt="alt text" width=500px>

paul-1 commented:
Yes the codec Digital volume is exposed, Those are the actual volume controls for each DAC.  Master is just a virtual device.
